bgpd/bgp_attr.c

```diff
@@ -2314,7 +2315,7 @@ bgp_attr_parse_ret_t bgp_attr_prefix_sid(struct bgp_attr_parser_args *args,
 	uint16_t length;
 	size_t headersz = sizeof(type) + sizeof(length);
 
-	while (STREAM_READABLE(peer->curr) > 0) {
+	while (STREAM_READABLE(peer->curr) > 0 && psid_attr_length) {
```

Comment:
A malformed packet with length value more than psid_attr_length would cause issue for this while loop. Like other if checks present in this while loop, it is better to have another break condition to break this loop.

Reply:
Can you add some more detail on what exactly the problem would be?

Reply:
I guess @srimohans want to say about following.

```diff
ecf9b7a087f9:~/frr.work [bgpd-fix-prefix-sid-parse-error] :) git diff
diff --git a/bgpd/bgp_attr.c b/bgpd/bgp_attr.c
index 3a89e07cf..cf7cb4071 100644
--- a/bgpd/bgp_attr.c
+++ b/bgpd/bgp_attr.c
@@ -2295,7 +2295,6 @@ bgp_attr_parse_ret_t bgp_attr_prefix_sid(struct bgp_attr_parser_args *args)
 {
        struct peer *const peer = args->peer;
        struct attr *const attr = args->attr;
-       uint16_t psid_attr_length = args->length;
        bgp_attr_parse_ret_t ret;

        attr->flag |= ATTR_FLAG_BIT(BGP_ATTR_PREFIX_SID);
@@ -2303,8 +2302,10 @@ bgp_attr_parse_ret_t bgp_attr_prefix_sid(struct bgp_attr_parser_args *args)
        uint8_t type;
        uint16_t length;
        size_t headersz = sizeof(type) + sizeof(length);
+       size_t psid_parsed_length = 0;

-       while (STREAM_READABLE(peer->curr) > 0 && psid_attr_length) {
+       while (STREAM_READABLE(peer->curr) > 0
+              && psid_parsed_length < args->length) {

                if (STREAM_READABLE(peer->curr) < headersz) {
                        flog_err(
@@ -2335,7 +2336,18 @@ bgp_attr_parse_ret_t bgp_attr_prefix_sid(struct bgp_attr_parser_args *args)
                if (ret != BGP_ATTR_PARSE_PROCEED)
                        return ret;

-               psid_attr_length -= length + headersz;
+               psid_parsed_length += length + headersz;
+
+               if (psid_parsed_length > args->length) {
+                       flog_err(
+                               EC_BGP_ATTR_LEN,
+                               "Malformed Prefix SID attribute - TLV overflow by attribute (need %zu"
+                               " for TLV length, have %zu overflowed in UPDATE)",
+                               length + headersz, psid_parsed_length - (length + headersz));
+                       return bgp_attr_malformed(
+                               args, BGP_NOTIFY_UPDATE_ATTR_LENG_ERR,
+                               args->total);
+               }
        }

        return BGP_ATTR_PARSE_PROCEED;
```

Reply:
Then, I'll fix my patch like an above change.

Reply:
I added more boundary check on https://github.com/FRRouting/frr/commit/38774fc5e6b8da326bbe3700d7d54498aedb154a#diff-d081a503fd40b967e697f8114c83ab58R2629-R2638

Reply:
> Can you add some more detail on what exactly the problem would be?

I missed to check quentin's question earlier. 

I meant, in the case where length is already more than psid_attr_length to start with,  then the math "psid_attr_length -= length + headersz;" can never be zero which results in infinte while loop. 

The modified code addresses this problem.